distributed/tests/test_asyncprocess.py

```diff
@@ -297,3 +300,124 @@ def test_terminate_after_stop():
+    # test will therefore be able to tell that this process has exited by
+    # reading children_alive.
+    _ = child_pipe
+
+    # Signal to parent process that this process has started and made it this
+    # far. This should cause the parent to exit rapidly after this statement.
+    worker_ready.close()
+
+    # The parent exiting should cause this process to os._exit from a monitor
+    # thread. This sleep should never return.
+    shorter_timeout = 2.5 # timeout shorter than that in the spawning this.
+    sleep(shorter_timeout)
+
+    # Unreachable if functioning correctly.
+    child_pipe.send("child chould have exited by now")
+
+
+def _parent_process(child_pipe):
+    """ Simulate starting an AsyncProcess and then dying.
+
+    The child_alive pipe is held open for as long as the child is alive, and can
+    be used to determine if it exited correctly. """
+    def parent_process_coroutine():
+        await_worker_start, worker_ready = multiprocessing.Pipe(duplex=False)
+
+        # Need to use "spawn" context so that the child_pipe file descriptor can
+        # be passed to the parent onwards. This does not occur with the
+        # forkserver multiprocessing context, somehow new file descriptors are
+        # created.
+        worker = AsyncProcess(target=_worker_process,
+                         args=(worker_ready, child_pipe,),
+                         mp_context=multiprocessing.get_context("spawn"))
+
+        yield worker.start()
+
+        # Close our reference to the worker_ready pipe, so that the child has
+        # the only one.
+        worker_ready.close()
+
+        # Wait for the child process to have started.
+        yield PipeIOStream(await_worker_start.fileno()).read_until_close()
+
+        # Exit immediately, without doing any process teardown (including atexit
+        # and 'finally:' blocks) as if by SIGKILL. This should cause
+        # worker_process to also exit.
+        os._exit(255)
+
+    with pristine_loop() as loop:
+        try:
+            loop.run_sync(gen.coroutine(parent_process_coroutine), timeout=10)
+        finally:
+            loop.stop()
+
+            raise RuntimeError("this should be unreachable due to os._exit")
+
+
+def test_asyncprocess_child_teardown_on_parent_exit():
+    """ Check that a child process started by AsyncProcess exits if its parent
+    exits.
+
+    The motivation is to ensure that if an AsyncProcess is created and the
+    creator process dies unexpectedly (e.g, via Out-of-memory SIGKILL), the
+    child process and resources held by it should not be leaked.
+
+    The child should monitor its parent and exit promptly if the parent exits.
+
+    [test process] -> [parent using AsyncProcess (dies)] -> [worker process]
+                 \                                          /
+                  \________ <--   child_pipe   <-- ________/
+    """
+    ctx = multiprocessing.get_context('spawn')
```

Comment:
I believe this is needed because of what I observed when I used the default `forkserver` - when I used the forkserver, the `child_pipe` passed into the `_parent_process` is not literally the same `fd` as the one that ends up being passed in to the `_worker`, so the signalling mechanism doesn't work. If you look closely at line 340 and also the diff to `AsyncProcess.__init__`, I had to also use the `spawn` method in `AsyncProcess`. This does unfortunately mean that this doesn't test the forking method ordinarily used, but I couldn't see a quick way to get it to work there. Any advice?